<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ резюме из HeadHunter

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

1. Прочитайте данные с помощью библиотеки Pandas. Совет: перед чтением обратите внимание на разделитель внутри файла. 

In [16]:
hh_df = pd.read_csv('data/dst-3.0_16_1_hh_database.csv', sep=';')
print(hh_df.shape)
display(hh_df.info())

(44744, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory

None

2. Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.

In [17]:
hh_df.head()

Пол, возраст           ЗП  \
0    Мужчина ,  39 лет , родился 27 ноября 1979   29000 руб.   
1     Мужчина ,  60 лет , родился 20 марта 1959   40000 руб.   
2  Женщина ,  36 лет , родилась 12 августа 1982   20000 руб.   
3      Мужчина ,  38 лет , родился 25 июня 1980  100000 руб.   
4     Женщина ,  26 лет , родилась 3 марта 1993  140000 руб.   

                           Ищет работу на должность:  \
0                            Системный администратор   
1                               Технический писатель   
2                                           Оператор   
3  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4                  Региональный менеджер по продажам   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев , не готов к переезду , готов к редким...   
2  Тверь , не готова к переезду , не готова к ком...   
3  Саратов , не готов к переезду , готов к редким...   
4  Москва , не готова к переезду , готова к коман...   

                                           Занятость  \
0  частичная занятость, проектная работа, полная ...   
1  частичная занятость, проектная работа, полная ...   
2                                   полная занятость   
3  частичная занятость, проектная работа, полная ...   
4                                   полная занятость   

                                              График  \
0  гибкий график, полный день, сменный график, ва...   
1  гибкий график, полный день, сменный график, уд...   
2                                        полный день   
3                    гибкий график, удаленная работа   
4                                        полный день   

                                         Опыт работы  \
0  Опыт работы 16 лет 10 месяцев  Август 2010 — п...   
1  Опыт работы 19 лет 5 месяцев  Январь 2000 — по...   
2  Опыт работы 10 лет 3 месяца  Октябрь 2004 — Де...   
3  Опыт работы 18 лет 9 месяцев  Август 2017 — Ап...   
4  Опыт работы 5 лет 7 месяцев  Региональный мене...   

  Последнее/нынешнее место работы  \
0         МАОУ "СОШ № 1 г.Немана"   
1    Временный трудовой коллектив   
2                    ПАО Сбербанк   
3                        OpenSoft   
4                        Мармелад   

                       Последняя/нынешняя должность  \
0                           Системный администратор   
1  Менеджер проекта, Аналитик, Технический писатель   
2                               Кассир-операционист   
3                               Инженер-программист   
4                              Менеджер по продажам   

                                   Образование и ВУЗ Обновление резюме  \
0  Неоконченное высшее образование 2000  Балтийск...  16.04.2019 15:59   
1  Высшее образование 1981  Военно-космическая ак...  12.04.2019 08:42   
2  Среднее специальное образование 2002 Профессио...  16.04.2019 08:35   
3  Высшее образование 2002  Саратовский государст...  08.04.2019 14:23   
4  Высшее образование 2015 Кгу Психологии и педаг...  22.04.2019 10:32   

                             Авто  
0  Имеется собственный автомобиль  
1                      Не указано  
2                      Не указано  
3                      Не указано  
4                      Не указано

In [18]:
hh_df.tail()

Пол, возраст           ЗП  \
44739    Мужчина ,  30 лет , родился 17 января 1989   50000 руб.   
44740      Мужчина ,  27 лет , родился 5 марта 1992   39000 руб.   
44741  Женщина ,  48 лет , родилась 26 декабря 1970   40000 руб.   
44742   Мужчина ,  24 года , родился 6 октября 1994   20000 руб.   
44743    Мужчина ,  38 лет , родился 25 апреля 1980  120000 руб.   

                               Ищет работу на должность:  \
44739  Финансист, аналитик, экономист, бухгалтер, мен...   
44740             Системный администратор, IT-специалист   
44741                         Аналитик данных, Математик   
44742                                   Контент-менеджер   
44743                               Руководитель проекта   

                            Город, переезд, командировки  \
44739  Тверь , готов к переезду (Москва, Химки) , гот...   
44740  Липецк , готов к переезду , готов к командировкам   
44741  Челябинск , готова к переезду , готова к редки...   
44742  Тамбов , не готов к переезду , не готов к кома...   
44743  Москва , не готов к переезду , не готов к кома...   

                                               Занятость  \
44739                                   полная занятость   
44740  проектная работа, частичная занятость, полная ...   
44741                                   полная занятость   
44742              частичная занятость, полная занятость   
44743                                   полная занятость   

                                                  График  \
44739                      полный день, удаленная работа   
44740  удаленная работа, гибкий график, полный день, ...   
44741                      полный день, удаленная работа   
44742                                   удаленная работа   
44743                                        полный день   

                                             Опыт работы  \
44739  Опыт работы 7 лет 7 месяцев  Финансист, аналит...   
44740  Опыт работы 7 лет  Системный администратор, IT...   
44741  Опыт работы 21 год 5 месяцев  Январь 1998 — по...   
44742  Опыт работы 3 года 10 месяцев  Контент-менедже...   
44743  Опыт работы 15 лет 10 месяцев  Руководитель пр...   

                         Последнее/нынешнее место работы  \
44739  ООО "IAS" (независимый участник объединения Ru...   
44740                                       ИП Пестрецов   
44741  ОАО «ЧМК», Исследовательско-Технологический Це...   
44742                                          IQ-Maxima   
44743                                  ПАО ГК ТНС энерго   

                    Последняя/нынешняя должность  \
44739                     Руководитель субгруппы   
44740                            Предприниматель   
44741                 Начальник группы аналитики   
44742                          Менеджер проектов   
44743  Руководитель отдела технической поддержки   

                                       Образование и ВУЗ Обновление резюме  \
44739  Высшее образование 2015  Московский гуманитарн...  22.04.2019 12:32   
44740  Высшее образование (Бакалавр) 2016  Воронежски...  22.04.2019 13:11   
44741  Высшее образование 2000  Южно-Уральский госуда...  09.04.2019 05:07   
44742  Высшее образование 2015  Тамбовский государств...  26.04.2019 14:25   
44743  Высшее образование 1997  Южно-Российский госуд...  05.07.2018 20:15   

                                 Авто  
44739                      Не указано  
44740                      Не указано  
44741                      Не указано  
44742  Имеется собственный автомобиль  
44743                      Не указано

3. Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

4. Обратите внимание на информацию о числе непустых значений.

In [19]:
display(hh_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

None

Первоначально, таблица с данными имеет 12 столбцов и 44744 записей, все данные имеют тип object.

В трёх столбцах из 12 -"Опыт работы", "Последнее/нынешнее место работы", "Последняя/нынешняя должность", имеются пропуски,все данные имеют тип object.

5. Выведите основную статистическую информацию о столбцах.

In [20]:
hh_df.describe()

Пол, возраст          ЗП  \
count                                           44744       44744   
unique                                          16003         690   
top     Мужчина ,  32 года , родился 17 сентября 1986  50000 руб.   
freq                                               18        4064   

       Ищет работу на должность:  \
count                      44744   
unique                     14929   
top      Системный администратор   
freq                        3099   

                             Город, переезд, командировки         Занятость  \
count                                               44744             44744   
unique                                              10063                38   
top     Москва , не готов к переезду , не готов к кома...  полная занятость   
freq                                                 1261             30026   

             График                                        Опыт работы  \
count         44744                                              44576   
unique           47                                              44413   
top     полный день  Опыт работы 10 лет 8 месяцев  Апрель 2018 — по...   
freq          22727                                                  3   

                          Последнее/нынешнее место работы  \
count                                               44743   
unique                                              30214   
top     Индивидуальное предпринимательство / частная п...   
freq                                                  935   

       Последняя/нынешняя должность  \
count                         44742   
unique                        16927   
top         Системный администратор   
freq                           2062   

                                        Образование и ВУЗ Обновление резюме  \
count                                               44744             44744   
unique                                              40148             18838   
top     Высшее образование 1987 Военный инженерный Кра...  07.05.2019 09:50   
freq                                                    4                25   

              Авто  
count        44744  
unique           2  
top     Не указано  
freq         32268

# <center>Преобразование данных

Начнём с простого — с признака «Образование и ВУЗ». Его текущий формат — это: <Уровень образования год выпуска ВУЗ специальность...>. Например:

Высшее образование 2016 Московский авиационный институт (Национальный исследовательский университет)
Неоконченное высшее образование 2000 Балтийская государственная академия рыбопромыслового флота
Нас будет интересовать только уровень образования. Причём, для простоты мы будем считать, что соискатель указывает свой максимальный уровень образования в первых трёх словах столбца “Образование и ВУЗ”.

Создайте с помощью функции-преобразования новый признак «Образование», который должен иметь четыре категории: «высшее», «неоконченное высшее», «среднее специальное» и «среднее».

Выполните преобразование, ответьте на контрольные вопросы и удалите признак «Образование и ВУЗ».

Совет первый
Обратите внимание на структуру текста в столбце «Образование и ВУЗ». Гарантируется, что текущий уровень образования соискателя всегда отражён в первых двух словах и начинается с заглавной буквы. Воспользуйтесь этим.

Совет второй
Проверяйте полученные категории, например с помощью метода unique()

Сколько соискателей имеет средний уровень образования (школьное образование)?

In [21]:
# группируем образование на 4 признака
def get_education(arg):
    arg = ' '.join(arg.split(' ')[:3])
    if 'Высшее' in arg:
        return 'высшее'
    elif 'Неоконченное высшее' in arg:
        return 'неоконченное высшее'
    elif 'Среднее специальное' in arg:
        return 'среднее специальное'
    elif 'Среднее образование' in arg:
        return 'среднее'

# создаём новый столбец 'Образование'   
hh_df['Образование'] = hh_df['Образование и ВУЗ'].apply(get_education)

# удаляем ненужный стобец 'Образование и ВУЗ'
hh_df = hh_df.drop('Образование и ВУЗ', axis=1)
hh_df['Образование'].value_counts()

высшее                 33863
среднее специальное     5765
неоконченное высшее     4557
среднее                  559
Name: Образование, dtype: int64

Теперь нас интересует столбец «Пол, возраст». Его текущий формат: <Пол , возраст, дата рождения>. Например:

Мужчина , 39 лет , родился 27 ноября 1979
Женщина , 21 год , родилась 13 января 2000
Как вы понимаете, нам необходимо выделить каждый параметр в отдельный столбец..

Создайте два новых признака «Пол» и «Возраст». При этом важно учесть:

Признак пола должен иметь два уникальных строковых значения: М — мужчина, Ж — женщина.
Признак возраста должен быть представлен целыми числами.
Выполните преобразование, ответьте на контрольные вопросы и удалите признак «Пол, возраст» из таблицы.
Совет
Обратите внимание на структуру текста в столбце, а точнее на то, как разделены параметры пола, возраста и даты рождения между собой — символом ' , '.

Гарантируется, что структура одинакова для всех строк в таблице. Вы можете воспользоваться этим.

1. Сколько процентов женских резюме представлено в наших данных? Ответ округлите до сотых.

2. Чему равен средний возраст соискателей? Ответ округлите до десятых.

In [22]:
#создаём функцию признака пола
def get_gender(arg):
    if 'Мужчина' in arg:
        return 'М'
    else:
        return 'Ж'

#создаём функцию признака возраста
def get_age(arg):
    arg_splited = arg.split(' ')
    year_words = ["год", "года", "лет"]
    for index, item in enumerate (arg_splited):
        if item in year_words:
            return int(arg_splited[index-1])

#создаём новый признак "Пол"
hh_df['Пол'] = hh_df['Пол, возраст'].apply(get_gender)

#создаём новый признак "Возраст"
hh_df['Возраст'] = hh_df['Пол, возраст'].apply(get_age)

#отвечаем на контрольные вопросы
print(round(hh_df['Пол'].value_counts(normalize=True)['Ж'] *100, 2))
print(round(hh_df['Возраст'].mean(), 2))

#удаляем признак "Пол, возраст"
hh_df = hh_df.drop('Пол, возраст', axis=1)

19.07
32.2


Следующим этапом преобразуем признак «Опыт работы». Его текущий формат: <Опыт работы: n лет m месяцев, периоды работы в различных компаниях…>.

Из столбца нам необходимо выделить общий опыт работы соискателя в месяцах, новый признак назовём «Опыт работы (месяц)».

Для начала обсудим условия решения задачи:

Во-первых, в данном признаке есть пропуски. Условимся, что, если мы встречаем пропуск, оставляем его как есть (функция-преобразование возвращает NaN).
Во-вторых, в данном признаке есть скрытые пропуски. Для некоторых соискателей в столбце стоит значение Не указано. Их тоже обозначим как NaN (функция-преобразование возвращает NaN)
В-третьих, нас не интересует информация, которая описывается после указания опыта работы (периоды работы в различных компаниях).

В-четвёртых, у нас есть проблема: опыт работы может быть представлен только в годах или только в месяцах. Например, можно встретить следующие варианты:

Опыт работы 3 года 2 месяца
Опыт работы 4 года
Опыт работы 11 месяцев
Учитывайте эту особенность в вашем коде.

В результате преобразования у вас должен получиться столбец, содержащий информацию о том, сколько месяцев проработал соискатель.

Выполните преобразование и удалите столбец «Опыт работы» из таблицы.

Чему равен медианный опыт работы (в месяцах) в нашей таблице?

In [23]:
#создаем функцию преобразования
def get_experience(arg):
    if arg is np.nan or arg == "Не указано":
        return None
    month_key_words = ['месяц', 'месяцев', 'месяца']
    year_key_words = ['год', 'лет', 'года']
    args_splited = arg.split(' ')[:7]
    month = 0
    year = 0
    for index, item in enumerate(args_splited):
        if item in month_key_words:
            month = args_splited[index-1]
        if item in year_key_words:
            year = args_splited[index-1]
    return int(year)*12 + int(month)

#создаем новый признак "Опыт работы(месяц)"
hh_df['Опыт работы (месяц)'] = hh_df['Опыт работы'].apply(get_experience)

#отвечаем на контрольные вопросы
print(round(hh_df['Опыт работы (месяц)'].median()))

#удаляем столбец "Опыт работы"
hh_df = hh_df.drop('Опыт работы', axis=1)

100


Хорошо идём! :) Следующий на очереди — признак «Город, переезд, командировки». Информация в нём представлена в следующем виде: <Город , (метро) , готовность к переезду (города для переезда) , готовность к командировкам>.

В скобках указаны необязательные параметры строки. Например, можно встретить следующие варианты:

Москва , не готов к переезду , готов к командировкам
Москва , м. Беломорская , не готов к переезду, не готов к командировкам
Воронеж , готов к переезду (Сочи, Москва, Санкт-Петербург) , готов к командировкам
Создадим отдельные признаки «Город», «Готовность к переезду», «Готовность к командировкам». При этом важно учесть:

⛏ Признак «Город» может содержать только четыре категории: «Москва», «Санкт-Петербург» и «город-миллионник» (их список ниже), остальные обозначьте как «другие».

Список городов-миллионников
million_cities = ['Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', 'Челябинск', 'Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж', 'Волгоград' ]
Информация о метро, рядом с которым проживает соискатель, нас не интересует.

⛏ Признак «Готовность к переезду» должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к переезду в признаке «Город, переезд, командировки». Например:

… , готов к переезду , …
… , не готова к переезду , …
… , готова к переезду (Москва, Санкт-Петербург, Ростов-на-Дону)
… , хочу переехать (США) , …
Нас интересует только сам факт возможности/желания переезда.

⛏ Признак «Готовность к командировкам» должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к командировкам в признаке «Город, переезд, командировки». Например:

… , готов к командировкам , …
… , готова к редким командировкам , …
… , не готов к командировкам , …
Нас интересует только сам факт готовности к командировке.

Ещё один важный момент: при выгрузке данных у некоторых соискателей «потерялась» информация о готовности к командировкам. Давайте по умолчанию будем считать, что такие соискатели не готовы к командировкам.

Выполните преобразования и удалите столбец «Город, переезд, командировки» из таблицы.

Совет
Обратите внимание на то, что структура текста может меняться в зависимости от указания ближайшего метро. Учтите это, если будете использовать порядок слов в своей программе.

1. Сколько процентов соискателей живут в Санкт-Петербурге? Ответ округлите до целого.

2. Сколько процентов соискателей готовы одновременно и к переездам, и к командировкам? Ответ округлите до целого.

In [24]:
#создаём функцию признака города
def get_city(arg):
    million_cities = ['Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', 'Челябинск', 'Омск', 'Самара',
                  'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж', 'Волгоград' ]
    city = arg.split(' , ')[0]
    if (city == 'Москва') or (city == 'Санкт-Петербург'):
        return city
    elif city in million_cities:
        return 'город миллионник'
    else:
        return 'другие'
    
#создаём функцию признака готовности к переезду
def get_removal(arg):
    if ('не готова к переезду' in arg) or ('не готов к переезду' in arg):
        return False
    elif 'хочу' in arg:
        return True
    else:
        return True
    
#создаём функицю признака готовности к командировкам
def get_business_trip(arg):
    if ('командировка' in arg):
        if ('не готова к командировкам' in arg) or ('не готов к командировкам' in arg):
            return False
        else:
            return True
    else:
        return False
    
#создаём новый признак "Город"
hh_df['Город'] = hh_df['Город, переезд, командировки'].apply(get_city)
#создаём новый признак "Готовность к перезду"
hh_df['Готовность к переезду'] = hh_df['Город, переезд, командировки'].apply(get_removal)
#создаём новый признак "Готовность к командировкам"
hh_df['Готовность к командировкам'] = hh_df['Город, переезд, командировки'].apply(get_business_trip)

#отвечаем на контрольные вопросы
print(round(hh_df['Город'].value_counts(normalize=True)['Санкт-Петербург'] *100, 2))
print(round(hh_df[hh_df['Готовность к переезду'] & hh_df['Готовность к командировкам']].shape[0] / hh_df.shape[0] *100))

#удаляем столбец "Город, переезд, командировки"
hh_df = hh_df.drop('Город, переезд, командировки', axis=1)

11.03
32


Рассмотрим поближе признаки «Занятость» и «График».

Сейчас признаки представляют собой набор категорий желаемой занятости (полная занятость, частичная занятость, проектная работа, волонтёрство, стажировка) и желаемого графика работы (полный день, сменный график, гибкий график, удалённая работа, вахтовый метод).

На сайте hh.ru соискатель может указывать различные комбинации данных категорий, например:

полная занятость, частичная занятость;
частичная занятость, проектная работа, волонтёрство;
полный день, удалённая работа;
вахтовый метод, гибкий график, удалённая работа, полная занятость.
Такой вариант признаков имеет множество различных комбинаций, а значит, множество уникальных значений, что мешает анализу. Нужно это исправить!

Давайте создадим признаки-«мигалки» для каждой категории: если категория присутствует в списке желаемых соискателем, то в столбце на месте строки рассматриваемого соискателя ставится True, иначе — False.

Такой метод преобразования категориальных признаков называется One Hot Encoding, его схема представлена на рисунке ниже:

![Схема](data/dst-3.0_16_3_1.png "Схема")

Выполните данное преобразование для признаков «Занятость» и «График», ответьте на контрольные вопросы, после чего удалите их из таблицы.

1. Сколько людей ищут проектную работу и волонтёрство (в обоих столбцах стоит True)?

2. Сколько людей хотят работать вахтовым методом и с гибким графиком (в обоих столбцах стоит True)?

In [25]:
#создаём списки из признаков "Занятость" и "График"
employments = ['полная занятость', 'частичная занятость', 'проектная работа', 'волонтерство', 'стажировка']
charts = ['полный день', 'сменный график', 'гибкий график', 'удаленная работа', 'вахтовый метод']

#при помощи цикла проходимся по данным спискам и создаём новые признаки
for employment, chart in zip(employments, charts):
    hh_df[employment] = hh_df['Занятость'].apply(lambda x: employment in x)
    hh_df[chart] = hh_df['График'].apply(lambda x: chart in x)

#отвечаем на контрольные вопросы
print(hh_df[hh_df['проектная работа'] & hh_df['волонтерство']].shape[0])
print(hh_df[hh_df['вахтовый метод'] & hh_df['гибкий график']].shape[0])

#удаляем признаки "Занятость" и "График"
hh_df = hh_df.drop('Занятость', axis=1)
hh_df = hh_df.drop('График', axis=1)

436
2311


Наконец, мы добрались до самого главного — признака заработной платы «ЗП». 

В чём наша проблема? — В том, что помимо желаемой заработной платы соискатель указывает валюту, в которой он хотел бы её получать, например:

30 000 руб.;
50 000 грн.;
550 USD.
Нам бы хотелось видеть заработную плату в единой валюте, например в рублях. Возникает вопрос: а где взять курс валют по отношению к рублю?

На самом деле язык Python имеет в арсенале огромное количество возможностей получения данной информации: от обращения к API Центробанка, до использования специальных библиотек, например [pycbrf](https://github.com/idlesign/pycbrf "pycbrf"). Однако в рамках нашего проекта эти темы мы рассматривать не будем.

Поэтому просто обратимся к профильным интернет-ресурсам для получения данных о курсе в виде текстовых файлов. Например,  [MDF.RU](https://mfd.ru/export/#Alias=false&Period=1&timeframeValue=1&timeframeDatePart=day&StartDate=04.10.2021&EndDate=04.10.2021&SaveFormat=0&SaveMode=0&FieldSeparator=%253b&DecimalSeparator=.&DateFormat=yyyyMMdd&TimeFormat=HHmmss&AddHeader=true&RecordFormat=0&Fill=false "MDF.RU"): данный ресурс позволяет удобно экспортировать данные о курсах различных валют и акций за указанные периоды в виде csv. Мы уже сделали выгрузку курсов валют, которые встречаются в наших данных за период с 29.12.2017 по 05.12.2019: скачать её вы можете  [здесь](https://lms.skillfactory.ru/assets/courseware/v1/15abf80f45a2f3e93c3274101b451c67/asset-v1:SkillFactory+DST-3.0+28FEB2021+type@asset+block/ExchangeRates.zip "здесь").

*Создайте новый DataFrame из скачанного файла*. В полученной таблице нас будут интересовать столбцы:

* currency — наименование валюты в ISO-кодировке;
* date — дата;
* proportion — пропорция;
* close — цена закрытия (последний зафиксированный курс валюты на указанный день).
Перед вами таблица соответствия наименований иностранных валют в наших данных и их общепринятых сокращений, которые представлены в файле с курсами валют.

*Пропорция* — это число, отражающее, скольким единицам валюты соответствует курс в таблице с курсами. Например для казахстанского тенге курс на 20.08.2019 составляет 17.197 руб. за 100 тенге, тогда итоговый курс равен: 17.197 / 100 = 0.17197 руб. за 1 тенге.

Воспользуйтесь этой информацией в ваших преобразованиях.

![Валюта](data/dst-3.0_16_3_2.png "Валюта")

Осталось только понять, откуда брать дату, по которой определяется курс. А вот же она — в признаке «Обновление резюме» содержится дата и время, когда соискатель выложил текущий вариант своего резюме. Нас интересует только дата, по ней мы и будем сопоставлять курсы валют.

Теперь у нас есть вся необходимая информация для того, чтобы создать признак «ЗП (руб)» — заработная плата в рублях.

Итак, давайте обсудим возможный алгоритм преобразования.

Подсказки
1
Обновить признак "Обновление резюме" из таблицы с резюме, переведя его в формат datetime и выделив из него дату. То есть в дате обновления резюме мы намеренно отсекаем время, оставляя только дату выставления резюме. В тот же формат привести признак date из таблицы с валютами. Обратите внимание на формат даты в исходных столбцах (день-месяц-год), учтите этот факт при преобразовании.

2
Выделить из столбца «ЗП» сумму желаемой заработной платы и наименование валюты, в которой она исчисляется. Наименование валюты перевести в стандарт ISO согласно таблице выше.

3
Присоединить к таблице с резюме таблицу с курсами по столбцам с датой и названием валюты (подумайте, какой тип объединения надо выбрать, чтобы в таблице с резюме сохранились данные о заработной плате, изначально представленной в рублях). Значение close для рубля заполнить единицей 1 (курс рубля самого к себе).

4
Умножить сумму желаемой заработной платы на присоединённый курс валюты (close) и разделить на пропорцию. Обратите внимание на пропуски после объединения в этих столбцах. Результат занести в новый столбец «ЗП (руб)».

Удалите исходный столбец заработной платы «ЗП» и все промежуточные столбцы, если вы их создавали.

Для выполнения этого задания у вас будет 5 попыток. После того как вы дадите правильный ответ ИЛИ истратите все попытки, вы можете нажать на кнопку «Показать ответ»: в конце решения вы увидите ссылку на эталонный ответ.

Чему равна желаемая медианная заработная плата соискателей в нашей таблице (в рублях)?

Ответ приведите в тысячах рублей и округлите до целого. Например, если у вас получилась медианная з/п 42 127, в качестве ответа укажите 42.

In [27]:
#создаём новый DataFrame из полученого файла
er_df = pd.read_csv('data/ExchangeRates.csv', sep=',')
# Удаляем "лишние" для наших преобразований столбцы из таблицы с курсами валют
er_df.drop(columns=['per', 'time', 'vol'], axis=1,  inplace=True)

#переводим признак "Обновление резюме" из таблицы с резюме в формат datetime 
hh_df['Обновление резюме'] = pd.to_datetime(hh_df['Обновление резюме'],
                                            dayfirst=True).dt.date
#переводим признак "date" из таблицы с валютами в формат datetime
er_df['date'] = pd.to_datetime(er_df['date'], dayfirst=True).dt.date

#выделяем сумму желаемой ЗП
def get_wages(arg):
    arg_list = float(arg.split(' ')[0])
    return arg_list
#наименование валюты переводим в стандарт ISO
def get_currency(arg):
    iso_std = {'грн':'UAH', 'USD':'USD',
               'EUR':'EUR', 'белруб':'BYN',
               'KGS':'KGS', 'сум':'UZS', 
               'AZN':'AZN', 'KZT':'KZT'}
    curr = arg.split(' ')[1].replace('.', '')
    if curr == 'руб':
        return 'RUB'
    else:
        return iso_std[curr]
    
#создаём временные признаки
hh_df['Валюта ISO'] = hh_df['ЗП'].apply(get_currency)
hh_df['ЗП_new'] = hh_df['ЗП'].apply(get_wages)

#присоединяем к таблице с резюме таблицу с курсом валют
merged = hh_df.merge(er_df, how='left',
                    left_on=['Обновление резюме', 'Валюта ISO'],
                    right_on=['date', 'currency'])

#заполним пропущенные значения в соответствии с заданием
merged['close'] = merged['close'].fillna(1)
merged['proportion'] = merged['proportion'].fillna(1)

# создадим новый столбец "ЗП (руб)"
hh_df["ЗП (руб)"] = merged["ЗП_new"]* merged['close'] / merged['proportion']

#отвечаем на контрольный вопрос
print(round(hh_df['ЗП (руб)'].median() / 1000))

#удаляем исходный столбец "ЗП" и все промежуточные
hh_df = hh_df.drop(['ЗП', 'Валюта ISO', 'ЗП_new'], axis=1)

59


Эталонный ответ вы можете посмотреть [здесь](https://drive.google.com/file/d/1j8hEX9ugaLYU_EBqjHdR9b8nMTsB0JBB/view?usp=sharing "эталон").

In [28]:
hh_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Ищет работу на должность:        44744 non-null  object 
 1   Последнее/нынешнее место работы  44743 non-null  object 
 2   Последняя/нынешняя должность     44742 non-null  object 
 3   Обновление резюме                44744 non-null  object 
 4   Авто                             44744 non-null  object 
 5   Образование                      44744 non-null  object 
 6   Пол                              44744 non-null  object 
 7   Возраст                          44744 non-null  int64  
 8   Опыт работы (месяц)              44574 non-null  float64
 9   Город                            44744 non-null  object 
 10  Готовность к переезду            44744 non-null  bool   
 11  Готовность к командировкам       44744 non-null  bool   
 12  полная занятость  